### PR TITLE
Changes

### PR DESCRIPTION
**Pull request by CaitSith2**
* Added to .gitignore and removed from repo all instances of exported_symbols.txt, patchData.json, and enemizerBasePatch.json
* Only copy those same 3 files only to places where they are needed for the build process.
* Cleanup now removes all instances of those same 3 files.
* assembly prebuild also removes all instances of those same 3 files.
* assembly postbuild builds patchData.json and exported_symbols.txt
* Asar errors are now shown if present in CompBin.
* enemizerBasePatch.json is now built by BasePatchGenerator postbuild rather than enemizer prebuild.
* Rom is now verified before it attempts to apply patchData.json to it.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -382,3 +382,6 @@ FodyWeavers.xsd
 /Setup/setup.7z
 Assembly/main.bin
 /add to release
+exported_symbols.txt
+patchData.json
+enemizerBasePatch.json
```
Comment (Ijwu) [inline .gitignore:387]: These entries will only ignore the files in the root directory of the repo. If you want to include all subdirectories check out the suggestion. I am unsure if this is intentional, so I figured I'd say something.

```suggestion
**/exported_symbols.txt
**/patchData.json
**/enemizerBasePatch.json
```

File: BasePatchGenerator/Program.cs [hunk 2]
```diff
@@ -3,36 +3,63 @@
...
-        static void Main(string[] args)
+	    private static void Main(string[] args)
         {
-            FileStream fs = new FileStream("zelda.sfc", FileMode.Open, FileAccess.Read);
-            byte[] rom_data = new byte[fs.Length];
+	        if (args.Length != 3)
+	        {
+				Console.WriteLine("Usage: BasePatchGenerator.exe [rom file] [patch data] [output file]");
+				return;
+	        }
```
Comment (Ijwu) [inline BasePatchGenerator/Program.cs:21]: The codebase seems to use 4-space indents, generally. I'd appreciate it if you converted away from tabs when commiting. Idc if you like to use tabs, it's all good, I just want to keep the code consistent.

Perhaps an .editorconfig might be a nice addition? If you think so, I'll make one.

File: EnemizerCLI.Core/Program.cs [hunk 3]
```diff
@@ -13,27 +13,27 @@ class Program
     {
         static void Main(string[] args)
         {
-            var options = new CommandLineOptions();
-            if (CommandLine.Parser.Default.ParseArgumentsStrict(args, options))
+	        var writer = new StringWriter();
```
Comment (Ijwu) [inline EnemizerCLI.Core/Program.cs:16]: Mixed tabs and spaces here.

File: EnemizerCLI.Core/Program.cs [hunk 4]
```diff
@@ -13,27 +13,27 @@ class Program
...
-				{
-                    stopwatch.Start();
-                    MakeEnemizerJsonPatch(options);
-                }
-                stopwatch.Stop();
-                Console.WriteLine($"Seed generated in: {stopwatch.Elapsed}");
-            }
-            else
+				Console.WriteLine(writer.ToString());
+				Environment.Exit(1);
```
Comment (Ijwu) [inline EnemizerCLI.Core/Program.cs:23]: Tabs here should probably be spaces.

This is the last message I'll put up about tabs, since that'll get tiring quick. A quick conversion over all the affected files should be quick and easy in VS. Let me know if it poses any problem.

File: EnemizerCLI.Core/Program.cs [hunk 5]
```diff
@@ -13,27 +13,27 @@ class Program
...
+            }).WithParsed(options =>
             {
-                Console.Write(GetUsage(options));
-            }
+				Stopwatch stopwatch = new Stopwatch();
+				stopwatch.Start();
```
Comment (Ijwu) [inline EnemizerCLI.Core/Program.cs:27]: You can shortcut creating and starting a new stopwatch.

```suggestion
            Stopwatch stopwatch = Stopwatch.StartNew();
```

File: EnemizerCLI.Core/Program.cs [hunk 6]
```diff
@@ -13,27 +13,27 @@ class Program
...
+
+	            if (options.BinaryMode)
+		            MakeEnemizerRom(options);
+	            else
+		            MakeEnemizerJsonPatch(options);
```
Comment (Ijwu) [inline EnemizerCLI.Core/Program.cs:32]: This is minor, but it's a code style I stick with always. Please use braces for all control flow statements, even one-liners.

File: bincomp/Program.cs [hunk 7]
```diff
@@ -28,10 +28,8 @@ private static void Main(string[] args)
 			Console.WriteLine($"Exported Symbols: {symbolsFilename}");
 
 			Console.WriteLine($"Cleaning up for a fresh build attempt.");
-			File.Delete(basePatchJson);
-			File.Delete(symbolsFilename);
 
-				// ReSharper disable InconsistentNaming
+			// ReSharper disable InconsistentNaming
 			// ReSharper disable JoinDeclarationAndInitializer
```
Comment (Ijwu) [inline bincomp/Program.cs:33]: I'm not using ReSharper at the moment, are you? If not, let's kill these.

File: bincomp/Program.cs [hunk 8]
```diff
@@ -117,6 +115,16 @@ private static void PrintUsage()
...
 		}
 
+		private static void CmdReadOutput(object sender, DataReceivedEventArgs e, AutoResetEvent outputWaitHandle, StringBuilder output)
+		{
+			if (e.Data == null)
+			{
+				outputWaitHandle.Set();
+				return;
+			}
+			output.AppendLine(e.Data);
```
Comment (Ijwu) [inline bincomp/Program.cs:125]: Asking out of ignorance/curiousity. Does the `AutoResetEvent` need to be reset at the end of the function (if `e.Data != null`)?

File: Assembly/Cleanup.bat [hunk 9]
```diff
@@ -0,0 +1,4 @@
+cd ..
+del /S patchData.json
+del /S exported_symbols.txt
+del /S enemizerBasePatch.json
```
Comment (Ijwu) [inline Assembly/Cleanup.bat:4]: This file matches the `Assembly` project pre-build. Is this file necessary?

File: bincomp/Program.cs (hunk 7 above)
Comment (CaitSith2) [reply bincomp/Program.cs:33]: I do use resharper

File: Assembly/Cleanup.bat (hunk 9 above)
Comment (CaitSith2) [reply Assembly/Cleanup.bat:4]: There is a call to this batch file.
In assembly.csproj, towards the end, are these lines.
    <Target Name="BeforeClean">
        <Exec Command="Cleanup.bat" ContinueOnError="true" />
    </Target>

This allows the files to be cleaned up, when you call clean project.

File: bincomp/Program.cs (hunk 8 above)
Comment (CaitSith2) [reply bincomp/Program.cs:125]: No actual need to Reset() the event.  It was started up in the Reset() state already.  Once there is no more data, it gets Set(), which allows the process to continue.

File: .gitignore (hunk 1 above)
Comment (CaitSith2) [reply .gitignore]: Commited and resolved

File: EnemizerCLI.Core/Program.cs (hunk 5 above)
Comment (CaitSith2) [reply EnemizerCLI.Core/Program.cs]: Committed and resolved

File: EnemizerCLI.Core/Program.cs (hunk 6 above)
Comment (CaitSith2) [reply EnemizerCLI.Core/Program.cs]: Okay, will keep that in mind with this project going forward.

File: BasePatchGenerator/Program.cs (hunk 2 above)
Comment (CaitSith2) [reply BasePatchGenerator/Program.cs:21]: Could totally use a .editorconfig.

Comment (Ijwu) [reply BasePatchGenerator/Program.cs]: I'll whip one up for future inclusion.

**Comment by Ijwu** [comment]
Thanks for your patience with me. I'll commit an .editorconfig then release a new version soon.

**Comment by Ijwu** [comment]
I brain farted and didn't merge. Doing that now. 🤦‍♂️
